### ISSUE TITLE
toLower crashes GHC for large inputs

### ISSUE THREAD

**Issue by tvh** [haskell/text#221]
It seems to want to allocate huge amounts of memory. It iven semms to overflow:

```
Prelude> import qualified Data.Text as T
Prelude T> :set -XOverloadedStrings 
Prelude T> T.toLower (T.replicate 200000 "0")
"<interactive>: internal error: Unable to commit -1604321280 bytes of memory
    (GHC version 8.2.2 for x86_64_unknown_linux)
    Please report this as a GHC bug:  http://www.haskell.org/ghc/reportabug
```

**Comment by tvh** [comment]
The first failing commit is cfb82784fb16e982960619af4187e83b07ac3fb2. In that commit the initial size of the buffer is increased, but only by a ffactor of 3. That should not couse a crash right?

**Comment by skogsbaer** [comment]
The bug also hit is in production. Any ideas how to fix it?

**Comment by hvr** [comment]
ping @bgamari -- since you authored the size-hints fix, do you see how this over-allocation (or maybe it's actually an integer underflow?) could occur?

**Comment by bgamari** [comment]
I will have a look once 8.6 alpha1 is out.

**Comment by hvr** [comment]
@bgamari as alpha1 was out some time ago already... did you have time to look at this?

**Comment by skogsbaer** [comment]
@bgamari any progress on this? I'm a bit worried because we have to work around the problem in production with a fix that is easy to forget or break.

**Comment by skogsbaer** [comment]
@bgamari any progress on this? I'm a bit worried because we have to work around the problem in production with a fix that is easy to forget or break.

**Comment by bgamari** [comment]
I'm looking at it now; sorry for the delay.

**Comment by nh2** [comment]
Can somebody confirm which `text` releases have this bug?

Apparently it's been introduced in https://github.com/haskell/text/commit/cfb82784fb16e982960619af4187e83b07ac3fb2#diff-3378cc573525bb476703cba5c67d5dbeR403 so is `text-1.2.3.0` the only version with this issue?

**Comment by nh2** [comment]
I have created https://ghc.haskell.org/trac/ghc/ticket/15581 for the fact that the number in `Unable to commit -1604321280 bytes of memory` is negative, that looks suspicious in general as well.

**Comment by hvr** [comment]
@nh2 yeah, it's definitely something I'd like to know too; i.e. where did the overflow originate exactly; i.e. where does the first signed negative interpretation of the amount occur

**Comment by osa1** [comment]
The negative memory problem is fixed in 8.6.1. As for why this program is failing, that's because `mmap()` is returning `MAP_FAILED`. I don't know if that's expected or not though..

**Comment by hvr** [comment]
@osa1 well, `mmap` returning MAP_FAILED w/ errno=ENOMEM is what I'd expect when trying to allocate more virtual memory than available/allowed via mmap(2)

Do you happen to know the GHC commit or GHC Trac ticket number which covered the fix that went into GHC 8.6.1?

**Comment by hvr** [comment]
Ah, I see... so it was merely an aesthetic bug
